cms/djangoapps/contentstore/outlines_backfill.py

```diff
@@ -0,0 +1,138 @@
+"""
+Code which updates and backfills course outline data.
+
+Uses a proxy model to enable a Django admin interface to trigger asynch
+tasks which backfill/recreate course outline data.
+"""
+import logging
+from django.contrib import admin
+from django.contrib.admin.helpers import ACTION_CHECKBOX_NAME
+
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from openedx.core.djangoapps.content.learning_sequences.api import key_supports_outlines
+
+from .tasks import update_outline_from_modulestore_task, update_multiple_outlines_from_modulestore_task
+
+
+# If an admin user requests the course outline backfilling of this number of courses -or- higher,
+# create a celery task which creates a celery tasks to backfill each course outline.
+TASK_TO_LAUNCH_TASKS_THRESHOLD = 20
+
+log = logging.getLogger(__name__)
+
+
+class ReadOnlyAdminMixin(object):
+    """
+    Disables all editing capabilities for the admin's model.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.list_display_links = None
+        self.readonly_fields = [f.name for f in self.model._meta.get_fields()]
+
+    def get_actions(self, request):
+        actions = super().get_actions(request)
+        if 'delete_selected' in actions:
+            del actions["delete_selected"]
+        return actions
+
+    def has_add_permission(self, request):
+        return False
+
+    def has_delete_permission(self, request, obj=None):  # pylint: disable=unused-argument
+        return False
+
+    def save_model(self, request, obj, form, change):
+        pass
+
+    def delete_model(self, request, obj):
+        pass
+
+    def save_related(self, request, form, formsets, change):
+        pass
+
+
+class CourseOutlineBackfill(CourseOverview):
+    """
+    Proxy model for CourseOverview.
+
+    Does *not* create/update/delete CourseOverview objects - only reads the objects.
+    Uses the course IDs of the CourseOverview objects to determine which course
+    outlines to re-build/backfill.
+    """
+    class Meta:
+        proxy = True
+
+    def __str__(self):
+        """Represent ourselves with the course key."""
+        return str(self.id)
+
+    @classmethod
+    def get_course_outline_ids(cls):
+        """
+        Returns all the CourseOverview object ids.
+        """
+        return cls.objects.values_list('id', flat=True)
+
+
+def backfill_course_outlines_subset(modeladmin, request, queryset):
+    """
+    Create a celery task to backfill a single course outline for each passed-in course key.
+
+    If the number of passed-in course keys is above a threshold, then instead create a celery task which
+    will then create a celery task to backfill a single course outline for each passed-in course key.
+    """
+    all_course_keys_qs = queryset.values_list('id', flat=True)
+    if len(all_course_keys_qs) >= TASK_TO_LAUNCH_TASKS_THRESHOLD:
```

Comment:
Nit: I wouldn't bother with the conditional. There's a practical limit to how many they can put in one of these (set by the number of results you decide to show on the Django admin list view), and I think it's fine to go through that queryset and queue a bunch of tasks in the web request.

Reply:
The `Backfill *all* course outlines` code path flows through this code also, which is why I included the threshold. Do you think I should instead just direct all numbers of backfills requested through the task-creating task? It would be a single code path and would be simpler in a sense.

Reply:
Nm - I'm changing this to "whatever can be selected on a single admin page" call path and "all" code path. Only "all" will use a task-creating task.

Reply:
I believe an admin user is able to select all records on all pages via the checkbox. I don't have enough courses in my devstack to trigger this behavior, but I can see it in the [prod waffle override admin](https://courses-internal.edx.org/admin/waffle_utils/waffleflagcourseoverridemodel/).
![image](https://user-images.githubusercontent.com/3628148/120336331-3c883180-c2c0-11eb-9c07-b40c43afe2d0.png)